docs/guides/VisitorPatterns.md

```diff
@@ -0,0 +1,227 @@
+# Visitor patterns: reading & writing with unknown column types
+
+ParquetSharp exposes a number of "visitor" interfaces that make it convenient to read or write columns when you don't know the concrete column types at compile time. These visitors let you write type-safe code that is invoked for the actual column element type at runtime.
+
+## ILogicalColumnWriterVisitor<TReturn>
+
+The `ILogicalColumnWriterVisitor<TReturn>` interface is invoked for logical writers (high-level typed writers). Use this when you need to write data to columns but don't know the column types at compile time.
+
+### Example: Generic column writer
+
+```csharp
+// A visitor that writes arrays of values to any column type
+sealed class GenericColumnWriter : ILogicalColumnWriterVisitor<bool>
+{
+    private readonly IDictionary<string, object> _valuesByColumn;
+
+    public GenericColumnWriter(IDictionary<string, object> valuesByColumn)
+    {
+        _valuesByColumn = valuesByColumn;
+    }
+
+    public bool OnLogicalColumnWriter<TValue>(LogicalColumnWriter<TValue> columnWriter)
+    {
+        // Look up values for this column name
+        if (!_valuesByColumn.TryGetValue(columnWriter.ColumnDescriptor.Path[0], out var raw))
```

Comment:
The `ColumnPath` type doesn't support indexing so this code won't compile. I think what you want is probably:
```suggestion
        if (!_valuesByColumn.TryGetValue(columnWriter.ColumnDescriptor.Path.ToDotString(), out var raw))
```

 Can you check all the examples compile and work?